File: 125-Valid Palindrome/PalindromeString.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    //1. lowercase and trims the spaces
    let str = s.toLowerCase().trim();

    /*
    2. removes all non-alphanumeric characters using a regular expression.
    The [^a-z0-9] pattern matches anything that is not a lowercase letter or a  digit (0-9)
    */
    let modifiedStr = str.replace(/[^a-z0-9]/g, '');

    //3. reverse the string
    let reversedStr = modifiedStr.split('').reverse().join('');

    // Compare the modified phrase(modifiedStr) with its reversedStr
    return modifiedStr === reversedStr;
};

console.log(isPalindrome("A man, a plan, a canal, Panama")); // Output: true
console.log(isPalindrome("race car")); // Output: true
console.log(isPalindrome("Hello, World!")); // Output: false

File: 125-Valid Palindrome/PalindromeString.java
class Solution {
    public boolean isPalindrome(String s) {
        // Convert the string to lowercase and remove non-alphanumeric characters
        String cleanString = s.toLowerCase().replaceAll("[^a-zA-Z0-9]", "");
        
        // Two pointers approach to check if cleanString is a palindrome
        int left = 0; // Initialize left pointer at the beginning of the string
        int right = cleanString.length() - 1; // Initialize right pointer at the end of the string
        
        while (left < right) {
            if (cleanString.charAt(left) != cleanString.charAt(right)) {
                return false; // Characters don't match, not a palindrome
            }
            left++; // Move left pointer to the right
            right--; // Move right pointer to the left
        }
        
        return true; // If all characters match, it's a palindrome
    }
    
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test cases
        String s1 = "A man, a plan, a canal: Panama";
        String s2 = "race a car";
        
        System.out.println("Is '" + s1 + "' a palindrome? " + solution.isPalindrome(s1));
        System.out.println("Is '" + s2 + "' a palindrome? " + solution.isPalindrome(s2));
    }
}
